### PR TITLE
More Nova forward ports

### PR DESCRIPTION
**Pull request by huitseeker**
Ports:
- last commit from https://github.com/microsoft/Nova/pull/203 (everything else was there already)
- https://github.com/microsoft/Nova/pull/254

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -542,24 +543,23 @@ where
...
 
     // check if the initial inputs match
-    if self.z0_primary != z0_primary || self.z0_secondary != z0_secondary {
-      return Err(NovaError::ProofVerifyError);
-    }
+    let is_inputs_not_match = self.z0_primary != z0_primary || self.z0_secondary != z0_secondary;
 
     // check if the (relaxed) R1CS instances have two public outputs
-    if self.l_u_secondary.X.len() != 2
+    let is_instance_has_two_outpus = self.l_u_secondary.X.len() != 2
```
Comment (mpenciak) [inline src/lib.rs:555]:
```suggestion
    let is_instance_has_two_outputs = self.l_u_secondary.X.len() != 2
```
nit: typo

**Review by mpenciak — COMMENTED** [review]
Small typo, but otherwise looks good!

**Review by adr1anh — APPROVED** [review]
LGTM

**Review by adr1anh — APPROVED** [review]
LGTM

**Pull request merged**